src/polydatum/dal.py

```diff
@@ -1,20 +1,236 @@
+        self.dal_method = None  # Not resolved yet
+
+
+class DalMethodError(Exception):
+    """
+    This class defines a DalMethodError that is raised
+    when a service/method cannot be resolved by the
+    DalMethodResolverMiddleware
+    """
+    def __init__(self, request: DalMethodRequest, path: Optional[Tuple[PathSegment]] = None):
+        self._request = request
+        self._path = path
+
+
+@dataclass
+class PathSegment:
+    """
+    This structure represents a segment of an attribute path.
+
+    For example, in this code:
+
+        dal.my_service.sub_service.other_service.method()
+
+    For the `dal` object, the entire attribute path would be:
+
+        my_service.sub_service.other_service.method
+
+    Which may be made up of PathSegments like:
+
+        PathSegment(name="my_service"), PathSegment(name="sub_service"), etc.
+    """
+    name: str
+
+    # A collection of meta properties.
+    # Example:
+    #   meta = {
+    #       "enabled": True
+    #   }
+    meta: dict
+
+    def __init__(self, name: str, **meta):
+        self.name = name
+        self.meta = meta
+
+
+class DalMethodRequester:
+    """
+    A utility class that defers access to DAL attributes.
+
+    Attribute access to the DAL is deferred so that the DAL can
+    run method middleware before and after calling a method.
+
+    Method middlewares may prevent method's from being called
+    by raising exceptions.
+    """
+
+    # Defining type here allows subclasses to easily provide another class
+    PathSegment = PathSegment
+
+    def __init__(self, handler: Callable, path: Tuple[PathSegment, ...]):
+        """
+        Args:
+            handler (Callable): A callable that handles calling the underlying
+                attribute by wrapping it so that method middleware can run before
+                and after calling a method
+                TODO: Consider using a custom type for the handler to represent
+                      or document the interface to that callable.
+            path (Tuple[PathSegment, ...]): A tuple of PathSegments representing
+                the current attribute path that has been requested.
+        """
+        self._handler = handler
+        self.path = path
+
+    def __getattr__(self, name: str) -> DalMethodRequester:
+        return self.__class__(self._handler, self.path + (self.PathSegment(name=name),))
+
+    def __call__(self, *args, **kwargs):
+        return self._handler(self.path, *args, **kwargs)
+
+
+class MethodMiddleware:
+    """
+    Base class used primarily for type checking and to provide
+    an interface for a method middleware.
+    """
+    def __call__(self, request: DalMethodRequest, handler: Callable):
+        pass
+
+
+def dal_resolver(ctx, dmr):
+    """
+    This function resolves a dal method call to its underlying
+    service
+    """
+    return resolve(DalMethodRequest(ctx=ctx, path=dmr.path, args=None, kwargs=None))
+
+
+def resolve(request: DalMethodRequest):
```

Comment:
Oh good point. However we had a need for this outside this class, that's why it's a function, BUT it was just for tests, so it's not a big deal to do something unique for tests.

Reply:
I thought of that too. It might be worth adding a `dal.resolve(path)` that can do it. But I don't think we want people to abuse it. Also, since resolving happens in middleware, it's hard to ask DAL to resolve without doing other middleware things.

I think that is fine. To test resolution, you can test DalResolverMiddleware directly without going through DM.

If you need a way to get the real service, you can do that by doing `dal._services["foo"]._services["bar"]`. Not pretty though.

Reply:
We removed all the MethodMiddleware stuff and now middlewares can be classes or functions. We also flipped requiring middleware classes to be types, instead of instances, and we instantiate them during init now, to be nice.